new/cookbook-segment-deprecation.md

```diff
@@ -0,0 +1,55 @@
+---
+RFC: unassigned
+Author: Lamont Granquist <lamont@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Title
+
+Cookbook Segment Deprecation
+
+## Motivation
+
+    As a Chef User/Developer
+    I want to be able to extend the structure of cookbooks easily,
+    in order to quickly adapt to new automation needs.
+
+## Specification
+
+This RFC will add a special `all_files` segment to cookbooks which will duplicate (initially)
+all of the existing cookbook manifest along with arbitrary files outside of the current scope of
+any of the cookbook segments.  The Chef Server will need to be patched in order to support this
+new 'segment' and accept it since the Chef Server currently does strict enforcement of file
+segments server-side.  This will also generate a version bump of the Chef Server protocol which
+will be used by servers to determine if the Chef Client supports this field or not.  Clients MAY
+implement code which determines the acceptable API version of the Server and then only send
+the new API requests to the server.  Clients MAY instead send requests which include both the
+old segment and new segment data.  The Server MUST detect the Client's desired API version and
```

Comment:
So I'm fairly certain that what I just wrote is broken.  If clients send up a cookbook version with 'all_files' set in addition to the other segments then on an old server it'll just barf.  That suggests to me that the RFC should really state that there is no 'mixed mode' and that servers will always know what the client supports so should send that format, and that clients MUST figure out if they're talking to an old server or not (we have no generic mechanism for this, so I can't suggest an implementation here, but for cookbook uploads it is simple and the client uploads all the sandbox files first and can determine the API version from those responses).  Then once the client determines the server API version it should send the appropriate requests.